core/lib/libmbedtls/src/tee_lmd_provider.c

```diff
@@ -358,15 +760,195 @@ TEE_Result crypto_acipher_rsaes_encrypt(uint32_t algo,
 					const uint8_t *src, size_t src_len,
 					uint8_t *dst, size_t *dst_len)
 {
-	return TEE_ERROR_NOT_IMPLEMENTED;
+	TEE_Result res = TEE_SUCCESS;
+	int lmd_res, lmd_padding;
+	size_t mod_size;
+	mbedtls_rsa_context rsa;
+	mbedtls_ctr_drbg_context ctr_drbg;
+	const mbedtls_pk_info_t *pk_info = NULL;
+	uint32_t md_algo = MBEDTLS_MD_NONE;
+
+	mbedtls_rsa_init(&rsa, 0, 0);
+
+	crypto_bignum_copy((void *)&rsa.E, key->e);
+	crypto_bignum_copy((void *)&rsa.N, key->n);
+
+	mod_size = crypto_bignum_num_bytes(key->n);
+	if (*dst_len < mod_size) {
+		*dst_len = mod_size;
+		res = TEE_ERROR_SHORT_BUFFER;
+		goto out;
+	}
+	*dst_len = mod_size;
+	rsa.len = mod_size;
+
+	if (algo == TEE_ALG_RSAES_PKCS1_V1_5)
+		lmd_padding = MBEDTLS_RSA_PKCS_V15;
+	else
+		lmd_padding = MBEDTLS_RSA_PKCS_V21;
+
+	pk_info = mbedtls_pk_info_from_type(MBEDTLS_PK_RSA);
+	if (pk_info == NULL) {
+		res = TEE_ERROR_NOT_SUPPORTED;
+		goto out;
+	}
+
+	/*
+	 * TEE_ALG_RSAES_PKCS1_V1_5 is invalid in hash. But its hash algo will
+	 * not be used in rsa, so skip it here.
+	 */
+	if (algo != TEE_ALG_RSAES_PKCS1_V1_5) {
+		md_algo = tee_algo_to_mbedtls_hash_algo(algo);
+		if (md_algo == MBEDTLS_MD_NONE) {
+			res = TEE_ERROR_NOT_SUPPORTED;
+			goto out;
+		}
+	}
+
+	mbedtls_rsa_set_padding(&rsa, lmd_padding, md_algo);
+
+	mbedtls_ctr_drbg_init(&ctr_drbg);
+	lmd_res = mbedtls_ctr_drbg_seed(&ctr_drbg, mbd_rand, NULL, NULL, 0);
+	if (lmd_res != 0) {
+		EMSG(" failed\n  ! mbedtls_ctr_drbg_seed ret is 0x%x\n",
+					-lmd_res);
+		res = TEE_ERROR_SECURITY;
+		goto out;
+	}
+
+	lmd_res = pk_info->encrypt_func(&rsa, src, src_len, dst,
+			dst_len, *dst_len,
+			mbedtls_ctr_drbg_random, &ctr_drbg);
+
+	switch (lmd_res) {
+	case MBEDTLS_ERR_RSA_INVALID_PADDING:
+	case MBEDTLS_ERR_PK_TYPE_MISMATCH:
+	case MBEDTLS_ERR_RSA_BAD_INPUT_DATA:
+		EMSG("encrypt_func() returned 0x%x\n", -lmd_res);
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto out;
+	case 0:
+		break;
+	default:
+		/* This will result in a panic */
+		res = TEE_ERROR_GENERIC;
+		goto out;
+	}
+	res = TEE_SUCCESS;
+out:
+	mbedtls_ctr_drbg_free(&ctr_drbg);
+	mbedtls_rsa_free(&rsa);
+	return res;
 }
 
 TEE_Result crypto_acipher_rsassa_sign(uint32_t algo, struct rsa_keypair *key,
 				      int salt_len, const uint8_t *msg,
 				      size_t msg_len, uint8_t *sig,
 				      size_t *sig_len)
 {
-	return TEE_ERROR_NOT_IMPLEMENTED;
+	TEE_Result res;
+	int lmd_res, lmd_padding;
+	size_t mod_size, hash_size;
+	mbedtls_rsa_context rsa;
+	mbedtls_ctr_drbg_context ctr_drbg;
+	const mbedtls_pk_info_t *pk_info = NULL;
+	uint32_t md_algo;
+
+	mbedtls_rsa_init(&rsa, 0, 0);
+
+	crypto_bignum_copy((void *)&rsa.E, key->e);
```

Comment:
Why does a second buffer holding the key need to be allocated?
Compare with `crypto_acipher_rsassa_sign()` in `core/lib/libtomcrypt/src/tee_ltc_provider.c`.

Reply:
It is because the struct of big number is different between mpa and mbedtls. 
The mpa bignum is this:
_typedef struct mpa_numbase_struct {
	mpa_asize_t alloc;
	mpa_usize_t size;
	mpa_word_t d[];
} mpa_num_base;_
the mbedTLS bignum is this:
_typedef struct
{
    int s;              /*!<  integer sign      */
    size_t n;           /*!<  total # of limbs  */
    mbedtls_mpi_uint *p;          /*!<  pointer to limbs  */
}
mbedtls_mpi;_
The data area is a pointer in mbedtls while it is a array in mpa. So if we assign it directly, The pointer(key) will be free when call mbedtls_rsa_free($rsa).

Reply:
The problem is that you double the amount of temporary memory needed to do these asymmetric operations.
Wouldn't it be enough to just copy the `mbedtls_mpi` structures? You'd need to be careful to not free those pointers at the end of `crypto_acipher_rsassa_sign()` though.

Reply:
Of course we can only copy the `mbedtls_mpi` structure, but we need place most of the context of  `mbedtls_rsa_free()` at the end of `crypto_acipher_rsassa_sign()`, like these:
`mbedtls_mpi_free( &ctx->Vi ); mbedtls_mpi_free( &ctx->Vf );
    mbedtls_mpi_free( &ctx->RQ ); mbedtls_mpi_free( &ctx->RP ); mbedtls_mpi_free( &ctx->RN );
    mbedtls_mpi_free( &ctx->QP ); mbedtls_mpi_free( &ctx->DQ ); mbedtls_mpi_free( &ctx->DP );
    mbedtls_mpi_free( &ctx->Q  ); mbedtls_mpi_free( &ctx->P  ); mbedtls_mpi_free( &ctx->D );
    mbedtls_mpi_free( &ctx->E  ); mbedtls_mpi_free( &ctx->N  );`
Do you think it make sense? If you think it is ok, I can push a new patch to do that.

Reply:
As I understand it only those `mbedtls_mpi` structures that we don't fill in with a bare `memcpy()` (`RN`, `RP`, `RQ`, `Vi`, `Vf`) need to be freed with `mbedtls_mpi_free()`.

Since the `E` to `QP` fields will be initialized with `memcpy()` or such I guess it would be OK to just clear them with a corresponding `memset()` or such before calling `mbedtls_rsa_free()` and we wouldn't need to care about how `mbedtls_rsa_context` is supposed to be cleaned up.

Reply:
Yes, this is a good solution. But the p to dq are optional here, so it need to add a conditional check if the p to dq need to `memset()`. Of course, I will change it in other functions.